### PR TITLE
Use a static cache for `PSVersionInfo.PSVersion` to avoid casting `SemanticVersion` to `Version` every time accessing that property

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Use a static cache for `PSVersionInfo.PSVersion` to avoid casting `SemanticVersion` to `Version` every time accessing that property

The work was initially aimed to avoid the ~40 ms CPU spent in the implicit conversion from `SemanticVersion` to `Version` during startup.
The implicit conversion would warp the `Version` instance to `PSObject` and add 2 `NoteProperty` to it (see the code snippet below). 
The ETS related operations is quite slow so the thought was it would be great if we can avoid it.

**However**, it turns out this is a problem **only for preview versions of PowerShell**, **not a problem for official releases**, because official releases won't have the `PreviewLabel` or `BuildLabel` and thus the ETS operations won't kick in during the implicit conversion from `SemanticVersion` to `Version`.

The only perf related change in this PR is to add a cache for `PSVersionInfo.PSVersion`, so it doesn't create a new `Version` instance every time it's accessed.
There is barely noticeable change for the warm startup scenario, but given `PSVesionInfo.PSVersion` is used a lot in the code base, the cache would potentially bring benefits by reducing allocation of `Version` instances in a relatively long running powershell process.

Changes:
1. add a cache for `PSVersionInfo.PSVersion`
2. use `s_psVersionTable` directly instead of calling `GetPSVersionTable()`
3. add `readonly` when applicable.

image

https://github.com/PowerShell/PowerShell/blob/f3a3922285cfd231c270d287c743b131b062443b/src/System.Management.Automation/engine/PSVersionInfo.cs#L588-L603

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 1]
```diff
@@ -173,23 +173,30 @@ public static Version PSVersion
         {
             get
             {
-                return (SemanticVersion)GetPSVersionTable()[PSVersionInfo.PSVersionName];
+                if (s_psVersionCache == null)
+                {
+                    // Ideally, we only do one conversion for 'PSVersion' and use the cached value afterwards.
+                    var version = (Version)(SemanticVersion)s_psVersionTable[PSVersionInfo.PSVersionName];
+                    Interlocked.CompareExchange(ref s_psVersionCache, version, null);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: It seems we always use the value. So we could initialize it in `static PSVersionInfo()`.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: That makes sense. Moved to the type initializer.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 2]
```diff
@@ -54,14 +52,15 @@ public class PSVersionInfo
...
-        private static SemanticVersion s_psV7Version;
+        private static readonly Version s_psV1Version = new Version(1, 0);
+        private static readonly Version s_psV2Version = new Version(2, 0);
+        private static readonly Version s_psV3Version = new Version(3, 0);
+        private static readonly Version s_psV4Version = new Version(4, 0);
+        private static readonly Version s_psV5Version = new Version(5, 0);
+        private static readonly Version s_psV51Version = new Version(5, 1, NTVerpVars.PRODUCTBUILD, NTVerpVars.PRODUCTBUILD_QFE);
+        private static readonly SemanticVersion s_psV6Version = new SemanticVersion(6, 0, 0, null, null);
+        private static readonly SemanticVersion s_psCurrentVersion;
+        private static readonly Version s_psVersionCache;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: If we use psVersionCache we could rename psCurrentVersion to psSemanticVersionCache.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: I will update the member names.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: Done.

**Comment by daxian-dbw** [comment]
@adityapatwardhan Can you please also review this? Thanks!

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 3]
```diff
@@ -54,14 +52,15 @@ public class PSVersionInfo
...
-        private static Version s_psV51Version = new Version(5, 1, NTVerpVars.PRODUCTBUILD, NTVerpVars.PRODUCTBUILD_QFE);
-        private static SemanticVersion s_psV6Version = new SemanticVersion(6, 0, 0, null, null);
-        private static SemanticVersion s_psV7Version;
+        private static readonly Version s_psV1Version = new Version(1, 0);
+        private static readonly Version s_psV2Version = new Version(2, 0);
+        private static readonly Version s_psV3Version = new Version(3, 0);
+        private static readonly Version s_psV4Version = new Version(4, 0);
+        private static readonly Version s_psV5Version = new Version(5, 0);
+        private static readonly Version s_psV51Version = new Version(5, 1, NTVerpVars.PRODUCTBUILD, NTVerpVars.PRODUCTBUILD_QFE);
+        private static readonly SemanticVersion s_psV6Version = new SemanticVersion(6, 0, 0, null, null);
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: nit: use named variables

**Review by adityapatwardhan — APPROVED** [review]

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: what do you mean by `named variables`?

Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: I meant named parameters. 

```suggestion
        private static readonly SemanticVersion s_psV6Version = new SemanticVersion(6, 0, 0, preReleaseLabel:null, buildLabel:null);
```

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs]: Updated.

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**